052-record-reconciled-versions.md

```diff
@@ -0,0 +1,160 @@
+
+For the `Kafka` Custom Resource it is often unclear when an upgrade has finished, whether all the brokers have rolled and to which version of the operator a cluster has been reconciled with.
+The following proposal puts forward a mechanism to make this clear in the `Kafka` custom resources status, making it easy to programmatically check for upgrade completion that could also be used/adopted in the upgrade system tests, simplifying the code and making for a better upgrade UX story.
+
+## Current situation
+
+Currently the only way to know if an upgrade has finished to completion is to check is all the zookeeper and kafka brokers have rolled.
+https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/systemtest/src/test/java/io/strimzi/systemtest/upgrade/KafkaUpgradeDowngradeST.java#L292
+Though this works, from a user standpoint it is far from ideal as it assumes a level of user knowledge about the cluster that may not be appropriate for all users.
+
+## Motivation
+
+This proposal allows a user to rely on the `Kafka` custom resources as their single source of truth, instead of watching/querying the individual kafka/zookeeper pods.
+This proposal lays out all the information we need and at what points in the reconcile. This information is already available within the Reconcile loop, so orchestrating and figuring out if a component has been fully upgraded will be propagated into the Custom Resource via status fields.
+
+## Proposal
+
+This proposal will cover the `Kafka` custom resource but could apply to other CR types that use `StrimziPodSets` such as `KafkaConnect` and `KafkaMirrorMaker2`, but could be extended to include other Custom Resources if there would be value in it.
+This proposal starts with the `Kafka` Custom Resource as the base example as it is the core component of Strimzi, and it deploys both `StrimziPodSet`s and `Deployment`s.
+
+This proposal assumes that all key value pairs mentioned will be writen to the `status` field as this is managed by the operator and can be added to without impacting end users.
+
+The following is the proposed new fields:
+
+### status.versions.strimziReconciled
+The `status.versions.strimziReconciled` field is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion.
+This would be the field a user or System Test would query against to check whether upgrade was complete.
+```
+status:
+  versions:
+    strimziReconciled: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.versions.strimziReconciled=0.35.0` would remain if a reconcilliation on `0.36.0` failed as the information that a prior reconcilliation passed could be useful to an end user.
+
+### status.versions.strimziManagedBy
```

Comment:
I have concerns about the usability of this. I do not think the difference between those two is really clear. If the reconciliation gets stuck at reconciling ZooKeeper or even before at the wrong version of Kafka for example, the `strimziManagedBy` will not really mean much as the operator does not really manage the cluster when it gets stuck right at the beginning. Similarly, if the reconciliation fails because of an issue with Kafka Exporter, the `strimziReconciled` might give the old version. But essentially everything other than the Kafka Exporter is being reconciled by the new version already.

So as an expert, you might understand these nuances. But then you should be able to figure it out from the resources and logs. And if you are just a regular user, you will not understand them and these two fields will be meaningless or confusing.

It also makes me a bit wonder of what are we trying to achieve:
* Is it the upgrade tracking? What exactly do you want to track? For example, if the upgrade jumps Kafka versions, the first successful reconciliation might not mean the end of it as the Kafka upgrade is multistage. Or if you want to track the 
* Is it trying to track which operator version is managing the cluster now? Maybe we should simplify it by having one version in the status and make it clear that the user needs to figure out from the logs / conditions how far did it got. But it would allow to see that this is now managed by a new version of the operator (successfully or not).

Reply:
So, in that case, maybe you could keep only one version field and update it at the successful reconciliation only. But for best understanding, maybe we should name it accordingly? `lastSuccessfulReconcilitationBy`? Or something similar to make it clear it talks about successful reconciliations? I think that would have some value, the name would make it easier to understand and with one version only it would be in general more understandable as there would not be the _How do they differ_ effect. So it is something I could live with for sure.

That said, my personal preference would be to keep the `managedBy` field and set it regardless of the reconciliation result. And your check that the upgrade completed would be that `managedBy` has new version and the Kafka CR is `Ready`. It would also let you see that `managedBy` has new version and the Kafka CR is **not** `Ready` in which case you basically know that the upgrade of the operator happened, but something went wrong when upgrading the operand. On the other hand, if the upgrade succeeded and then it failed for an unrelated reason, while you went away from the keyboard, you will not know that the reconciliation passed before. So, I guess it has different points of view. But as I said, I would be fine with the above as well - just sharing for consideration.

---

The Kafka version was not mentioned so far. I think it would be useful to have it - especially since the `.spec.kafka.version` is not mandatory. I guess the challenge is:
* When do you update it? With respect to the reconciliation progress - after the reconciliation succeeds or at some point when KafkaReconciler succeeds?
* Is this something for Kafka CR? Should we consider having the Kafka version defined per node pool? It was not discussed in detail. But for example some users do upgrade through rolling out new brokers with new Kafka versions, reassigning the partitions and then delete the old brokers. So we should consider how likely it is that we will support this in the future and if yes, what do we do with this field. Otherwise I think it would be useful.

Reply:
This has given me a lot of food for thought, especially regarding the KafkaNodePools upgrade option.
I think we could also add something like a `status.versions.kafka` to the `KafkaNodePool` CR as well, and this would be reflected also in the Kafka CR, since the KafkaReconciler should 'know' about which version of Kafka it is switching to through the Node Pools.

I think I will edit the proposal to now cover 3 status fields (names not yet confirmed):
```
status.versions.lastSuccessfulReconcilitationBy
status.versions.managedBy
status.versions.kafka
```
These will be included in the `Kafka` CR.

And then a user combining and checking `status.versions.lastSuccessfulReconcilitationBy` , `status.versions.kafka` and `status.phase==Ready` would cover a great deal of upgrade scenarios being 'finished'/successful.

the `status.versions.managedBy` field being updated would assure a user that an upgrade was atleast starting, but the value of this field can still be debated.

Reply:
But why mix both status.versions.lastSuccessfulReconcilitationBy and status.versions.managedBy together? Wouldn't one be enough? Two introduce the confusion of what is the difference. As for the node pools, I would not add anything there right now as they are not reconciled separately. And the Strimzi operator will be always the same as for the Kafka resource. But we need to consider if the Kafka version in the Kafka CR is future proof or not.

Reply:
I will rewrite, combining status.versions.lastSuccessfulReconcilitationBy and status.versions.managedBy
If I come up with a compelling reason in the flow to add it back in, I will.
Behaviour for `lastSuccessfulReconcilitationBy` will be, update on success, leave it as-is if present on failure.

Reply:
It is not about compelling reasons. It is about usability. You can add 10 different versions. But then it will be completely confusing what they mean, what is the difference, which one they need to care about and so on. So I think you should pick one.